### ISSUE TITLE
Elliptic curve testing (using GMP-ECM's B2 and curve counts)

### ISSUE THREAD

**Issue by grandpascorpion** [cartazio/arithmoi#21]
Hi,

I noticed that GMP-ECM (and other code bases apparently) uses different B2 and curve count values:
http://www.mersennewiki.org/index.php/Elliptic_Curve_Method

Given it's GMP-ECM, I would assume these figures are the best known.

(The B2 and curve counts are the 3rd and 4th members of each tuplet in testParms list in Math/NumberTheory/Primes/Factorisation/Montgomery.hs)   

The arithmoi code has an entry for 12 (digit factors).   The GMP-ECM tables do not.  Is the 12 entry really needed? Oddly, there is "no" optimal number of curves for 15 in the GMP-ECM.  Not sure how to interpret that.

Do you think it's worth changing the arithmoi code?

**Comment by Bodigrim** [comment]
Thank you for your question. 

Currently `arithmoi` uses B1 and curve counts from https://www.alpertron.com.ar/ECM.HTM for 15..50 digits. Alas, I do not know the source of parameters for 55..60 digits.

The entry for 12 digits is not really important, because it is weird to factorise such small numbers via ECM. IMHO for the same reason it makes no harm.

Choice of B2 is not a dogma, usually it is chosen to spend the same magnitude of time in stage 2 as in stage 1. GMP-ECM source code says:
```c
  /* set second stage bound B2: when using polynomial multiplication of
     complexity n^alpha, stage 2 has complexity about B2^(alpha/2), and
     we want stage 2 to take about half of stage 1, thus we choose
     B2 = (c*B1)^(2/alpha). Experimentally, c=1/4 seems to work well.
     For Toom-Cook 3, this gives alpha=log(5)/log(3), and B2 ~ (c*B1)^1.365.
     For Toom-Cook 4, this gives alpha=log(7)/log(4), and B2 ~ (c*B1)^1.424. */
```
This means that the value of `c` is implementation dependent. In order to change current values of B2 with a good reason we need to write benchmarks to determine `c`. Pull requests are welcome :) We use `criterion` to write benchmarks, I'll be glad to help with technical details.

It seems reasonable to choose curve count in the same way as GMP-ECM does (`print_expcurves` in `ecm.c`). Again, pull requests are very welcome.

**Comment by grandpascorpion** [comment]
Thanks for the detail

**Comment by Bodigrim** [comment]
Closed in #65.
